### PR TITLE
fix: create2 deploy to non-zero balance addres

### PR DESCRIPTION
**Pull request by lightsing**

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/util/common_gadget.rs [hunk 1]
```diff
@@ -374,32 +374,37 @@ impl<F: Field, const N_ADDENDS: usize, const INCREASE: bool>
 /// setting it's code_hash = EMPTY_HASH.   The receiver account is also created
 /// unconditionally if must_create is true.  This gadget is used in BeginTx.
 #[derive(Clone, Debug)]
-pub(crate) struct TransferWithGasFeeGadget<F> {
-    sender_sub_fee: UpdateBalanceGadget<F, 2, false>,
+pub(crate) struct TransferMayWithGasFeeGadget<F, SenderSubFee> {
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/util/common_gadget.rs:377]: The comment above are outdated.  After line 369.

**Comment by lispc** [comment]
\b run testool

**Comment by lispc** [comment]
{keccak_}code_hash_previous is free cell, unconstrained.   Not sound?  Can be fixed later.

**Comment by lispc** [comment]
[common_gadget.rs](https://github.com/scroll-tech/zkevm-circuits/pull/727/files/6bf182a47bf55766e63eee084cd7d88eff283283#diff-08d58ebc59ab6c11bf7c049e859441eed2319a03feb3fdffaddaba73f8a8c829) The comment above are outdated. After line 369.

**Comment by lispc** [comment]
we can refactor coinbase creation too https://github.com/scroll-tech/zkevm-circuits/pull/588  i have another refactor idea:

```
struct TransferMayWithGasFeeGadgetInner<F> {
    sender_sub_value: UpdateBalanceGadget<F, 2, false>,
    receiver: UpdateBalanceGadget<F, 2, true>,
    receiver_exists: Expression<F>,
    must_create: Expression<F>,
```

move "sender_sub_value" out of this gadget, and rename it to 'TransferTo' gadget.
And create a:

```
struct TransferFromGadget<const WITH_FEE: bool> {
    sender_sub_fee: UpdateBalanceGadget<F, 2, false>, // if WITH_FEE == false, do nothing in construct/assign method of this gadget?
    sender_sub_value: UpdateBalanceGadget<F, 2, false>,
```

then we can use only the new TransferTo for coinbase refund

**Comment by lispc** [comment]
btw i think the assignment codes for TransferXXXGadget in different opcodes is very similar too. Error prone.  It will be good if we can rewrite it like `fn TransferXXXGadget::assign_from_rws(block: &Block<Fr>, step: &Step, rw_offset: &mut usize)`.

**Comment by lightsing** [comment]
> we can refactor coinbase creation too #588 i have another refactor idea:
> 
> ```
> struct TransferMayWithGasFeeGadgetInner<F> {
>     sender_sub_value: UpdateBalanceGadget<F, 2, false>,
>     receiver: UpdateBalanceGadget<F, 2, true>,
>     receiver_exists: Expression<F>,
>     must_create: Expression<F>,
> ```
> 
> move "sender_sub_value" out of this gadget, and rename it to 'TransferTo' gadget. And create a:
> 
> ```
> struct TransferFromGadget<const WITH_FEE: bool> {
>     sender_sub_fee: UpdateBalanceGadget<F, 2, false>, // if WITH_FEE == false, do nothing in construct/assign method of this gadget?
>     sender_sub_value: UpdateBalanceGadget<F, 2, false>,
> ```
> 
> then we can use only the new TransferTo for coinbase refund

what do you mean by "do nothing in construct/assign method of this gadget", the gadget is still required for construct this struct. or you want using generic like current approach replace it with an ZST?

**Pull request merged**